### PR TITLE
[keycloak] Add autoscaling support and KUBE_PING documentation

### PR DESCRIPTION
**Pull request by cablespaghetti**
We've been implementing autoscaling for our Keycloak environment as well as switching to KUBE_PING after having issues with scaling and DNS_PING. Here are the changes we had to make to get it working.

Previous experiments with autoscaling keycloak did not go well but the scaling policies in Kubernetes 1.18 as well as KUBE_PING seem to be the answer :smile: 

https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-configurable-scaling-behavior

### CODE REVIEW

File: charts/keycloak/values.yaml [hunk 1]
```diff
@@ -405,6 +405,32 @@ prometheusRule:
   #   labels:
   #     severity: warning
 
+autoscaling:
+  # If `true`, a autoscaling/v2beta2 HorizontalPodAutoscaler resource is created (requires Kubernetes 1.18 or above)
+  # Autoscaling seems to be most reliable when using KUBE_PING service discovery (see README for details)
+  # This disables the `replicas` field in the StatefulSet
+  create: false
```
Comment (unguiculus) [inline charts/keycloak/values.yaml:412]: I'd rather use `enabled` instead of `create` here.

File: charts/keycloak/values.yaml [hunk 2]
```diff
@@ -405,6 +405,32 @@ prometheusRule:
...
+  minReplicas: 3
+  maxReplicas: 10
+  # The metrics to use for scaling
+  metrics:
+  - type: Resource
+    resource:
+      name: cpu
+      target:
+        type: Utilization
+        averageUtilization: 80
```
Comment (unguiculus) [inline charts/keycloak/values.yaml:423]:
```suggestion
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 80
```

File: charts/keycloak/values.yaml [hunk 3]
```diff
@@ -405,6 +405,32 @@ prometheusRule:
...
+  # The scaling policy to use. This will scale up quickly but only scale down a single Pod per 5 minutes.
+  # This is important because caches are usually only replicated to 2 Pods and if one of those Pods is terminated this will give the cluster time to recover.
+  behavior:
+    scaleDown:
+      stabilizationWindowSeconds: 300
+      policies:
+      - type: Pods
+        value: 1
+        periodSeconds: 300
```
Comment (unguiculus) [inline charts/keycloak/values.yaml:432]:
```suggestion
        - type: Pods
          value: 1
          periodSeconds: 300
```

File: charts/keycloak/templates/hpa.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{{- if .Values.autoscaling.create }}
+apiVersion: autoscaling/v2beta2
+kind: HorizontalPodAutoscaler
+metadata:
+  name: {{ include "keycloak.fullname" . }}
```
Comment (unguiculus) [inline charts/keycloak/templates/hpa.yaml:5]: Add labels as for all other resources.

File: charts/keycloak/README.md [hunk 5]
```diff
@@ -321,6 +329,58 @@ extraEnv: |
     value: "2"
 ```
 
+#### KUBE_PING Service Discovery
+
+Recent versions of Keycloak include a new Kuberntes Native [KUBE_PING](https://github.com/jgroups-extras/jgroups-kubernetes) service discovery protocol. This requires a little more configuration than DNS_PING but it is not difficult through the Helm Chart.
```
Comment (unguiculus) [inline charts/keycloak/README.md:334]:
```suggestion
Recent versions of Keycloak include a new Kubernetes native [KUBE_PING](https://github.com/jgroups-extras/jgroups-kubernetes) service discovery protocol.
This requires a little more configuration than DNS_PING but can easily be achieved with the Helm chart.
```

File: charts/keycloak/README.md [hunk 6]
```diff
@@ -321,6 +329,58 @@ extraEnv: |
...
+      fieldRef:
+        apiVersion: v1
+        fieldPath: metadata.namespace
+  - name: CACHE_OWNERS_COUNT
+    value: "2"
+  - name: CACHE_OWNERS_AUTH_SESSIONS_COUNT
+    value: "2"
+```
+
+However the Keycloak Pods must also be able to `get` and `list` Pods in the namespace which can be configured as follows:
```
Comment (unguiculus) [inline charts/keycloak/README.md:353]:
```suggestion
However, the Keycloak Pods must also get RBAC permissions to `get` and `list` Pods in the namespace which can be configured as follows:
```

File: charts/keycloak/README.md [hunk 7]
```diff
@@ -321,6 +329,58 @@ extraEnv: |
...
+
+```yaml
+rbac:
+  create: true
+  rules:
+  RBAC rules for KUBE_PING
+    - apiGroups:
```
Comment (unguiculus) [inline charts/keycloak/README.md:360]: This line must be removed or commented out
```suggestion
  rules:
    - apiGroups:
```

File: charts/keycloak/README.md [hunk 8]
```diff
@@ -321,6 +329,58 @@ extraEnv: |
...
+      resources:
+        - pods
+      verbs:
+        - get
+        - list
+```
+
+#### Autoscaling
+
+Due to the caches in Keycloak only replicating to a few nodes (2 in the example configuration above) and the limited controls around autoscaling built into Kubernetes it has historically been problematic to autoscale Keycloak. However in Kubernetes 1.18 [additional controls were introduced](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-configurable-scaling-behavior) which make it possible to scale down in a more controlled manner.
```
Comment (unguiculus) [inline charts/keycloak/README.md:371]:
```suggestion
Due to the caches in Keycloak only replicating to a few nodes (two in the example configuration above) and the limited controls around autoscaling built into Kubernetes, it has historically been problematic to autoscale Keycloak.
However, in Kubernetes 1.18 [additional controls were introduced](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#support-for-configurable-scaling-behavior) which make it possible to scale down in a more controlled manner.
```

File: charts/keycloak/README.md [hunk 9]
```diff
@@ -321,6 +329,58 @@ extraEnv: |
...
+
+The example autoscaling configuration in the values file scales from 3 up to a maximum of 10 Pods using CPU utilization as the metric. Scaling up is done as quickly as required but scaling down is done at a maximum rate of one Pod per 5 minutes.
```
Comment (unguiculus) [inline charts/keycloak/README.md:373]:
```suggestion
The example autoscaling configuration in the values file scales from three up to a maximum of ten Pods using CPU utilization as the metric. Scaling up is done as quickly as required but scaling down is done at a maximum rate of one Pod per five minutes.
```

**Review by unguiculus — CHANGES_REQUESTED** [review]
That's a nice addition. Thanks 🎉

**Comment by cablespaghetti** [comment]
I plead innocence on the pipeline failure :innocent:

**Review by unguiculus — APPROVED** [review]

**Pull request merged**